### PR TITLE
Adjusted MCLMC

### PR DESCRIPTION
**Pull request by reubenharry**
A change is also made to `partially_refresh_momentum`, which allows (adjusted) MCHMC to be the limit of MCLMC with infinite $L$. This is convenient since the user may or not want partial refreshment (in the unadjusted case, we find that it is always beneficial, but don't know yet in the adjusted case)

Addresses #676.

Checklist:

 - [x] *If I add a new sampler*, there is an issue discussing it already;
 - [x] We should be able to understand what the PR does from its title only;
 - [x] There is a high-level description of the changes;
 - [x] There are links to *all* the relevant issues, discussions and PRs;
 - [x] The branch is rebased on the latest `main` commit;
 - [x] Commit messages follow these [guidelines](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html);
 - [x] The code respects the current naming conventions;
 - [x] Docstrings follow the [numpy style guide](https://numpydoc.readthedocs.io/en/latest/format.html)
 - [x] `pre-commit` is installed and configured on your machine, and you ran it before opening the PR;
 - [x] There are tests covering the changes;
 - [x] The doc is up-to-date;
 - [x] *If I add a new sampler** I added/updated related [examples](https://github.com/blackjax-devs/blackjax/tree/main/examples)

### CODE REVIEW

**Comment by reubenharry** [comment]
Any advice for how to improve the code coverage here? I'm not entirely sure how to interpret the report.

File: blackjax/adaptation/mclmc_adaptation.py [hunk 1]
```diff
@@ -298,6 +302,310 @@ def step(state, key):
     return adaptation_L
 
 
+Lratio_lowerbound = 0.0
+Lratio_upperbound = 2.0
+
+
+def adjusted_mclmc_find_L_and_step_size(
```
Comment (junpenglao) [inline blackjax/adaptation/mclmc_adaptation.py:309]: The code added here has quite a bit of duplication with the non adjusted version:
```
mclmc_find_L_and_step_size
make_L_step_size_adaptation
make_adaptation_L
```

I suggest you create a new file, that way when we are comparing the implementation and trying to abstract the 2 it is a bit easier. WDYT?

Comment (reubenharry) [reply blackjax/adaptation/mclmc_adaptation.py]: OK, I split it out. There is similar code, but the details of the respective functions differ in important ways. It's true that with effort, one could probably abstract the code a little, but I'm not yet convinced that the gain would be huge. At a later date, we're planning to do some refactoring of the tuning (https://github.com/blackjax-devs/blackjax/issues/677) and that might be a good point to revisit this.

**Comment by reubenharry** [comment]
I think this is ready to review again @junpenglao

**Comment by reubenharry** [comment]
@junpenglao Now ready to review! I will extend the sampling book's MCLMC section with the adjusted option shortly.

File: blackjax/adaptation/adjusted_mclmc_adaptation.py [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+
+    if frac_tune3 != 0:
+        for i in range(num_windows):
+            part2_key = jax.random.fold_in(part2_key, i)
+            part2_key1, part2_key2 = jax.random.split(part2_key, 2)
+
+            state, params = adjusted_mclmc_make_adaptation_L(
+                mclmc_kernel,
+                frac=frac_tune3,
+                Lfactor=0.5,
```
Comment (junpenglao) [inline blackjax/adaptation/adjusted_mclmc_adaptation.py:103]: I dont quite understand some of the parameter choice, but i trust that you have test them or they are mostly arbitory.

File: blackjax/adaptation/adjusted_mclmc_adaptation.py [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+                params.step_size,
+                step_size_max,
+                info.energy,
+            )
+
+            log_step_size, log_step_size_avg, step, avg_error, mu = update_da(
+                adaptive_state, info.acceptance_rate
+            )
+
+            adaptive_state = DualAveragingAdaptationState(
```
Comment (junpenglao) [inline blackjax/adaptation/adjusted_mclmc_adaptation.py:171]: seems `mask * new_param + (1 - mask) * old_param` is the common pattern, worth to turn it into a lambda function

File: blackjax/adaptation/adjusted_mclmc_adaptation.py [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+                params = params._replace(
+                    step_size=mask * step_size + (1 - mask) * params.step_size,
+                )
+
+            else:
+                params = params._replace(
+                    step_size=mask * step_size + (1 - mask) * params.step_size,
+                    L=mask * (params.L * (step_size / params.step_size))
+                    + (1 - mask) * params.L,
+                )
```
Comment (junpenglao) [inline blackjax/adaptation/adjusted_mclmc_adaptation.py:205]:
```suggestion
            params = params._replace(
                step_size=mask * step_size + (1 - mask) * params.step_size,
            )
            if not fix_L:
                params = params._replace(
                    L=mask * (params.L * (step_size / params.step_size))
                    + (1 - mask) * params.L,
                )
```

**Review by junpenglao — APPROVED** [review]
Minor nits only.

File: blackjax/adaptation/adjusted_mclmc_adaptation.py (hunk 2 above)
Comment (reubenharry) [reply blackjax/adaptation/adjusted_mclmc_adaptation.py:103]: Actually this is a good reminder: I updated the tuning so this part is off by default, since it doesn't in general yield an improvement - this is reflected in the sampling book section too. (I want it to be available though, so have not removed it from the code)

**Review by junpenglao — APPROVED** [review]
Thank you!
